File: silver/9012/python/main.py

if (__name__ == "__main__"):
    num = int(input())

    for i in range(num):
        stack = list()
        text = input()
        
        for data in text:
            if ((not stack) and data == ')'):
                stack = [1]
                break
            if (data == '('):
                stack.append(data)
            elif (data == ')' and stack[-1] == '('):
                stack.pop(-1)
            else:
                stack.append(data)
        print("YES" if (not stack) else "NO")

File: silver/9012/c++/main.cpp
#include <iostream>
#include <string>
#include <stack>

int main()
{
    std::string text;

    int num;
    std::cin >> num;
    std::cin.ignore();

    for (int i = 0; i < num; i++) {
        std::stack<char> array; 

        std::getline(std::cin, text);
        for (char data : text) {
            if (array.size() == 0 && data == ')') {
                array.push('n');
                break;
            }

            if (data == '(') {
                array.push(data);
            } else if (data == ')' && array.top() == '(') {
                array.pop();
            } else {
                array.push(data);
            }
        }

        std::cout << (array.size() == 0 ? "YES" : "NO") << '\n';
    }

    return 0;
}